api/inc/rpc_gateway_exports.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __UVISOR_API_RPC_GATEWAY_EXPORTS_H__
+#define __UVISOR_API_RPC_GATEWAY_EXPORTS_H__
+
+#include "api/inc/uvisor_exports.h"
+#include <stdint.h>
+
+/* udf imm16
+ * UDF - ARMv7M ARM section A7.7.191
+ * 111 1;0 111;1111; <imm4>; 1 01 0; <imm12> (Encoding T2)
+ */
+#define UDF_OPCODE(imm16) \
+    ((uint32_t) (0xA000F7F0U | (((uint32_t) (imm16) & 0xFFFU) << 16U) | (((uint32_t) (imm16) & 0xF000U) >> 12U)))
+
+/** RPC gateway magics
+ *
+ * The following magics are used to verify an RPC gateway structure. The magics are
+ * chosen to be one of the explicitly undefined Thumb-2 instructions.
+ */
+/* TODO Unify all sources of magic (for register gateway, rpc gateway, and
+ * everybody else) */
+#if defined(__thumb__) && defined(__thumb2__)
+#define UVISOR_RPC_GATEWAY_MAGIC_ASYNC UDF_OPCODE(0x07C2)
+#define UVISOR_RPC_GATEWAY_MAGIC_SYNC  UDF_OPCODE(0x07C3)
+#else
+#error "Unsupported instruction set. The ARM Thumb-2 instruction set must be supported."
+#endif   /* __thumb__ && __thumb2__ */
+
+
+/** RPC gateway structure
+ *
+ * This struct is packed because we must ensure that the `ldr_pc` field has no
+ * padding before itself and will be located at a valid instruction location,
+ * and that the `function` field is at a pre-determined offset from the
+ * `ldr_pc` field.
+ */
+typedef struct {
+    uint32_t ldr_pc;
```

Comment:
Just to make sure that this is tested: The function that is loaded into the `pc` becomes a direct jump and never returns to the gateway, since `lr` is never updated. Besides, even if that function returned, it would then step over literal pool data (the metadata).

Has it been tested/verified in ASM that the destination function (the one that posts the RPC) correctly returns to the original caller (the caller of the gateway)?